### ISSUE TITLE
BIG CHANGE! increase storage battery capacity and electric motor epower drain

### ISSUE THREAD

**Issue by mlangsdorf** [CleverRaven/Cataclysm-DDA#25139]
#### Background: Weak storage batteries ####
A vehicle power unit is 373 watts over 6 seconds.
A storage battery contains 40,000 vehicle power units.
That works out to 4.14 kW-hrs.

Current real-world storage batteries range from 6.5 kW-hrs (2019 Toyota Prius) to 60+ kW-hrs for a Tesla roadster.

So our storage batteries are somewhere between 1/5th and 1/12th the size they should be.

#### Background: Impossibly efficient electric motors ####
A standard electric motor consumes 5 kW of electric power to produce 37.3 kW of motor power.
A large electric motor consumes 40 kW of electric power to produce 149 kW of motor power.
An enhanced electric motor consumes 60 kW of electric power to produce 186.5 kW of motor power.
(by way of comparison, a truck alternator consumes 2.2 kW of motor power to produce 1.3 kW of electric power.)

So our electric motors are 7 to 3x as efficient as it is physically possible to be.

#### Two problems, one solution ####
Make the electric motors 80% efficient (possibly 85% for the enhanced), so their numbers look like:

Type | epower drain | motor power
---- | ------------ | ----
Electric Motor | 46.625 kW | 37.3 kW
Large Electric Motor | 186.kW | 149.2 kW
Enhanced Electric Motor | 215 kW |186.5 kW

Increase storage battery size to 300,000 vehicle power units (roughly 30 kW-hrs).

Net effect is that electric vehicles drain their batteries roughly 3x or 4x as fast, but the battery is 7.5x as large, so the effective range is doubled. But since the recharge rate isn't increased, it takes 7x as long to recharge a full discharged battery (via solar power or whatever), and the range from one day's charge is reduced by a factor of 3 or 4.

It will be a substantial nerf to pure solar vehicles. Hybrid electric vehicles become much more attractive.

On the plus side, stationary craft rigs get a slight boost, because a fully charged battery takes 7x as long to discharge. A single mini-fridge should take roughly 30 days to discharge a single fully charged storage battery, so powering a survivor's house through the winter should be much easier.

#### Additional Changes ####
Add hydraulic lifts as furniture to gas stations and garages, so that there's another path to swapping out storage batteries.

Possibly add furniture/terrain rechargers to labs, the refugee center, and other locations that are established as having independent power supplies.

#### Alternate Solutions ####
This would be a substantial change to the game. It's not strictly required, but the fact that electric motors are 300+% efficient is deeply weird.

**Comment by ymber** [comment]
This would definitely be an improvement. Vehicle power and electric motors are way out right now.

Changing vehicle power to use an appropriate standard unit would make it easier to set good values for vehicle components. Watt hours would work fine for anything inside the scope of the vehicle system.

**Comment by lethargicchii** [comment]
I may be newish to the community and this game, but, I wholeheartedly agree with the @mlangsdorf logic. Exact numbers will probably need testing and such for balance, but, this seems like a good idea

**Comment by CoroNaut** [comment]
Solar panels being upped as well or no? Also the alternators and generator might be affected?

**Comment by mlangsdorf** [comment]
No change to solar panels, alternators, generators, or the combustion engines. Those numbers are good enough. Standard automotive batteries may be a small boost.

**Comment by tyrael93** [comment]
It's the future
Everything's more efficient

**Comment by mlangsdorf** [comment]
Solar panels already have two upgrades that improve their efficiency, and I will reintroduce rotating solar panels from blazemod at some point. But quantum solar panels, iirc, are already close to 99% efficient, which is pretty much the upper bounds of reasonable.

**Comment by kevingranade** [comment]
https://en.m.wikipedia.org/wiki/Shockley%E2%80%93Queisser_limit

Tl;dr conventional solar cells at around 33%, advanced panels at around 40%, theoretical advanced panels at about 75%, quantum panels theoretically at 100%.

Basically, if anything, everything but quantum panels would drop in output, quantum would probably stay where it is.

**Comment by mlangsdorf** [comment]
The current ratio of quantum to upgraded to standard is 6:2:1 (300 epower, 100 epower, 50 epower). It could be 4:2:1 (300 epower, 150 epower, 75 epower) if that 24% efficiency number is correct.

But this isn't a discussion about solar panels: I'm not changing them for this issue. This proposal is solely about electric motor efficiency and storage battery capacity.

**Comment by sfsworms** [comment]
According to the item browser, a storage battery is 22.86 kg, whereas the 85 kWh tesla battery pack is 540 kg. (6,85 kg/kWh) The Prius battery pack seems to be around 80 kg. (~12 kg/kWh) (Which brings into question why we currently need a crane to install a 22kg battery pack, but that's another debate). 

So if our storage battery was Tesla-efficient, it'd store 3.57 kWh. If right now it stores 4.14, it's about right.

I think upping the electric engines' power consumption is fair, but then we should either bump up the wiehgt of a storage battery a lot, or add a new "Large Storage Battery" and add it to the various electric vehicles. (And maybe remove the lift requirement from the current storage battery)

**Comment by Night-Pryanik** [comment]
> According to the item browser, a storage battery is 22.86 kg

I don't know where from item browser takes info, but according to the `data\json\items\vehicle\battery.json` storage battery weights a whooping 150 kg, that's why it requires crane to install.

**Comment by sfsworms** [comment]
Uh, I think I was looking at the stable, not the development for the item browser, my bad. 30 kWh for a 150 kg pack is perfectly reasonable.

**Comment by Lorith** [comment]
I don't like the idea of my batteries draining even faster, but it seems like a good change to me.  As long as solar panels don't get nerfed without also reducing the whole raining 80% of the time, it seems fair to me.  I have always been annoyed at how rapidly batteries discharge even without driving nonstop.

I had never looked into the numbers, but it did always seem like electric engines were a bit too strong.

**Comment by cake-pie** [comment]
Hang on.

> A vehicle power unit is 373 watts over 6 seconds.

This is well documented in the code.

The conversion comes from 0.5hp ≈ 373W
One vehicle power unit is 0.5hp over 1 tick
One vehicle epower unit is 1W over 1 tick = 6J = 1/600 Wh


vehicle epower generation/consumption is straightforward:
A 7.5kW generator has `"epower" : 7500` and produces 45 kJ per tick
A floodlight has `"epower": -1500` and consumes 9 kJ per tick
A small electric motor has `"epower": -1000` and consumes 6 kJ per tick

Batteries, however...

> A storage battery contains 40,000 vehicle power units.

This struck me as rather strange, so I [checked](https://github.com/CleverRaven/Cataclysm-DDA/blob/673ac458d903f7dab5947ff42f60fbfa8937327f/src/vehicle.cpp#L2754-L2760) -- vehicle battery charge is indeed given in vehicle power units.

Why are they not simply in epower?

This means that one charge of vehicle battery is equivalent to 2238J or 0.621667 Wh, how odd.

Side note: lead acid batteries changed in #17183 (a6d817aeed9abb17dd62909278aefcfaf5f09444) probably have incorrect values as a result of misunderstanding the units.

> That works out to 4.14 kW-hrs.

Don't we need to multiply by (1 tick = 6s)?

I think it should be:
```
40,000 vehicle power = 14920 kW over 1 tick
                     = 89520 kJ
                     = 24.8667 kWh
```

**Comment by mlangsdorf** [comment]
I'm not exactly sure why vehicle power units are used for storage battery charges, and it does cause problems.

Power is generated and consumed in 6 second chunks. A minifridge (which draws 250 W, which is reasonably for conventional minifridges) drains a storage battery in 4 days, but if you work out the math, it should drain it in about 30 days. That was the discussion that original prompted the "storage batteries are too small" revelation.

**Comment by mlangsdorf** [comment]
Yes, I want to move vehicle power calculations to watts and energy calculations to joules.

Okay, you've convinced me I did the math wrong. So now we're just going to increase electric motor power consumption by a factor of 3x to 7x without a corresponding increase in battery capacity, and people are going to howl.

**Comment by cake-pie** [comment]
> Yes, I want to move vehicle power calculations to watts and energy calculations to joules.

Yeah, we're going to need that, because I found inconsistent behavior when I dug further into the code. :sweat:
It'll need fixed, more on that in a bit. First, re: proposed motor changes:

Enhanced motor at 215 kW (same as Tesla Roadster) looks to be a sensible choice for highest powered motor.

As for efficiency, from [wikipedia](https://en.wikipedia.org/wiki/Electric_car#Energy_efficiency)
> electric vehicles have on-board efficiency of over 90%, when counted against stored chemical energy, or around 80%, when counted against required energy to recharge.

80% is measured against mains electricity consumed, and includes losses from charging system, such as converting mains AC to DC for the battery, etc.

In-game power generation using solar panels and ICE-powered alternators/generator are lossy; jumper cables also cause power loss.  So, perhaps we could consider letting electric motors be closer to 90% efficiency so as not to double-penalize the player for charging costs.

And the nitty-gritty:

> A standard electric motor consumes 5 kW of electric power to produce 37.3 kW of motor power.

Well, only if we're sure that [`"power": 100`](https://github.com/CleverRaven/Cataclysm-DDA/blob/02e0a7f6c196f4fa4f701aec315a5ad0473e0f78/data/json/vehicleparts/motor.json#L46) indeed translates to 50 horsepower in the code that calculates acceleration, moves the vehicle, etc. Could you please verify?  (I haven't looked at the movement code in-depth yet.)

The reason I don't take that for granted is the inconsistency I uncovered when looking at the code that handles fuel consumption:
- [here](https://github.com/CleverRaven/Cataclysm-DDA/blob/02e0a7f6c196f4fa4f701aec315a5ad0473e0f78/src/vehicle_move.cpp#L189) is the function call that should consume fuel proportionate to acceleration when thrust is applied

- fuel consumption accounts for fractional charges here:
https://github.com/CleverRaven/Cataclysm-DDA/blob/02e0a7f6c196f4fa4f701aec315a5ad0473e0f78/src/vehicle.cpp#L2631-L2641
  - L2631 suggests that fuel consumption of all fuel types is handled in kilojoules; this contradicts epower consumption code that used 1 battery charge = 3 horsepower-seconds :confused:
  - not sure about L2638, expected it to be
   `fuel_remainder[ ft ] = drain_energy( ft, amnt_precise ) - amnt_precise ;`
- [this](https://github.com/CleverRaven/Cataclysm-DDA/blob/02e0a7f6c196f4fa4f701aec315a5ad0473e0f78/src/vehicle_part.cpp#L262) is where energy consumption is ultimately translated into charges of fuel consumed

`JSON_INFO.md` states:
```
"fuel" : {
    energy": 34.2,    // battery charges per mL of fuel. batteries have energy 1
                      // is also MJ/L from https://en.wikipedia.org/wiki/Energy_density
                      // assumes stacksize 250 per volume 1 (250mL). Multiply
                      // by 250 / stacksize * volume for other stack sizes and volumes
```
which suggests that batteries are intended to have energy 1 (= 1MJ/L)
Energy content of all other fuels is specified in terms of this benchmark.

Now, **IF** batteries had `stack_size` 250, then the above would be true, and also agree with `vehicle.cpp#L2631` that one battery charge is 1kJ.  However, batteries are implicitly defined with `stack_size` 100 due to [`count` 100](https://github.com/CleverRaven/Cataclysm-DDA/blob/02e0a7f6c196f4fa4f701aec315a5ad0473e0f78/data/json/items/ammo.json#L16). And just to muddy things further, thanks to fractional volumes for AMMO types being rounded (#22028) the actual behavior in game is that 100 charges are 200mL i.e. `stack_size` 125  :dizzy_face:

If 1 charge = 1kJ then 40,000 vpu = 11.11kWh -- certainly on the lower side of things

Anyhow, bottom line is: it's a bit of a mess, all of that will need to be made consistent.

Considering that the [item definition for battery](https://github.com/CleverRaven/Cataclysm-DDA/blob/02e0a7f6c196f4fa4f701aec315a5ad0473e0f78/data/json/items/ammo.json#L2-L20) is used not only for vehicle batteries but also dry cells and other stored charge in various tools, it needs to be appropriate for lower-power items as well. AFAICT those specify `turns_per_charge` and are handled in `item::process_tool()`. As a point of reference, currently a flashlight [lasts 2 min per charge of battery](https://github.com/CleverRaven/Cataclysm-DDA/blob/02e0a7f6c196f4fa4f701aec315a5ad0473e0f78/data/json/items/tool/lighting.json#L77). Assuming a 2-cell (3V) flashlight with a 0.5W bulb: 60J in 2 min -- nowhere near 1kJ as per the markdown docs or 3 horsepower-seconds as per epower code.

---

> now we're just going to increase electric motor power consumption by a factor of 3x to 7x without a corresponding increase in battery capacity, and people are going to howl.

Well, two factors from the code discussed so far:
- common always-on appliances fridge/freezer draw way too much power
  - perceived as batteries being underpowered
- fuel consumption code uses 1 charge = 1kJ
  - indeed leaves batteries a bit underpowered

Both can be fixed.

But something else that plays into this is the perception that electric vehicles are underpowered compared to combustion (#20411). And I think I've found the cause for that:

[wikipedia](https://en.wikipedia.org/wiki/Electric_car#Energy_efficiency):
> Gasoline engines effectively use only 15% of the fuel energy content to move the vehicle or to power accessories, and diesel engines can reach on-board efficiency of 20%

[This here](https://github.com/CleverRaven/Cataclysm-DDA/blob/02e0a7f6c196f4fa4f701aec315a5ad0473e0f78/src/vehicle.cpp#L2595) looks like combustion engines are converting fuel to energy at 100% efficiency. That means ICE is overpowered, having fuel efficiency at 500%~670% what it ought to be. Electric motors being at 300%~700% efficiency just happened to slightly compensate.

Likewise, while alternators are lossy, the combustion engine used to power them are overpowered, thus making solar seem much weaker by comparison.

---

> [make] hybrid electric vehicles [...] more attractive.

Here's a related idea: currently engines are binary on/off. On a hybrid EV, turning both motor and engine on leads to both of them contributing motive power, but with a [further penalty in efficiency](https://github.com/CleverRaven/Cataclysm-DDA/blob/02e0a7f6c196f4fa4f701aec315a5ad0473e0f78/src/vehicle.cpp#L2195) -- not exactly ideal, so I personally tend to switch between pure electric and pure combustion in such situations.

What we could do is have an additional mode available to combustion engines on hybrid EVs, where they only power the alternator and not the wheels -- i.e. the range extender role. With careful balance tweaks, this could be made into an attractive choice for fuel economy.

---

Sorry about massive wall of text, one thing led to another, etc.
Would appreciate my findings being double checked.

**Comment by mlangsdorf** [comment]
Glancing through your logic, I think you're correct in your assumptions, and yes, the vehicle code is a mess and 100 nominal HP does not produce the acceleration or velocity that 100 HP should.

I'm now thinking a better approach is to hold off on this - if both combustion engines and motors are stupidly efficient, they're stupidly efficient in the same ways.
Instead, I (or other people who want to work on this crud) need to first figure out, and then update, the vehicle movement code so that the numbers look about right.

Basically, we need to decide on a bunch of reference vehicles and get performance data on them, then model similar vehicles in CDDA and tune things until the results are sensible:

Real World Vehicle | Curb Weight (lbs) | Engine | Engine HP | Max Speed (mph)
------------------- | -------- | --------- | -------- | ----
Porsche 911 GT3 Sports car | 3150 | 4.0L i6 | ~500 | 195
Mazda6 Sedan | 3600 | 2.5L i4 | 184 | 105?
Ford Fiesta Subcompact | 2340 | 1.6 i4 | 180 | 136
Yamama VX1900A Motorcylce | 725 | 1.8L v2 | 81 | 90?

We could probably use better sample vehicles than this.

**Comment by cake-pie** [comment]
> 100 nominal HP does not produce the acceleration or velocity that 100 HP should.

Hm, this should impact both combustion and EVs equally, right?
In which case I think it should be fine if we take start taking steps to alleviate some of the found problems now, as well as lay a bit of groundwork for later.

Stuff like:
- fix fridge/freezer power consumption (#25536)
- streamline and unify code for supporting fuel efficiency for both engines and motors
  - only json changes will be needed when we are ready to stop engines/motors being stupidly efficient
  - I have something in mind already, is next on my todo list.
- change epower consumption code to treat 1 charge = 1kJ for now
  - 3 horsepower-seconds is just awkward
  - makes consistent with fuel consumption code
  - worry about redefining the amount of energy per 1 battery charge later

> figure out, and then update, the vehicle movement code so that the numbers look about right.

sgtm. :+1:
Will assist as able.

> model similar vehicles in CDDA and tune things until the results are sensible

One concern: are vehicle parts stats sane enough that we can expect CDDA models of reference vehicles to behave "sensibly"? I've just increased sheet metal by 6× as part of review of steel resources #25210, don't know about the status of other vehicle parts.

**Comment by AbsalomAchitophel526** [comment]
Right, I've been doing some tests with the stock (spawned using debug menu) scooter vs electric scooter.  Here's my data.

Going at 10 mph (assuming a fully charged battery, straight road course, and nude driver) with cruise control, an electric scooter has 10 minutes of drive time, or 7.5 overmap tiles.  Incidentally, the stock scooter has no method of recharging its batteries.  Going to 20 mph, however, has 8 minutes of drive time but nearly 12 tiles of distance traveled due to coasting.  I will define "coasting" as un-powered travelling that still achieves a velocity of greater than or equal to 10 mph.  The trend continues as you start with higher speeds with cruise control, ending with nearly 35 tiles at 80 mph.

Without cruise control, you can achieve much greater distances.  To make a long story short, the most efficient way to use the electric scooter is manually going at ~60 mph.  You will travel nearly 55~ tiles.

Incidentally, the regular, gasoline powered scooter, has a max safe speed of 22 mph... but can travel the same max distance, 55~ tiles, with only 12% fuel spent.

**Comment by mlangsdorf** [comment]
Closing this issue for now. There needs to be a general rework of vehicle speed mechanics (more on the line of this http://www.asawicki.info/Mirror/Car%20Physics%20for%20Games/Car%20Physics%20for%20Games.html). When a 100 HP motor is providing sensible speeds for a small car or motorcycle, I will look into a general rebalance of consumption rates,
